src/runtime/objmodel.cpp

```diff
@@ -905,43 +960,57 @@ Box* typeLookup(BoxedClass* cls, const std::string& attr, GetattrRewriteArgs* re
     if (rewrite_args) {
         assert(!rewrite_args->out_success);
 
-        RewriterVar* obj_saved = rewrite_args->obj;
-
-        auto _mro = cls->tp_mro;
-        assert(_mro->cls == tuple_cls);
-        BoxedTuple* mro = static_cast<BoxedTuple*>(_mro);
-
-        // Guarding approach:
-        // Guard on the value of the tp_mro slot, which should be a tuple and thus be
-        // immutable.  Then we don't have to figure out the guards to emit that check
-        // the individual mro entries.
-        // We can probably move this guard to after we call getattr() on the given cls.
-        //
-        // TODO this can fail if we replace the mro with another mro that lives in the same
-        // address.
-        obj_saved->addAttrGuard(offsetof(BoxedClass, tp_mro), (intptr_t)mro);
-
-        for (auto base : *mro) {
-            rewrite_args->out_success = false;
-            if (base == cls) {
-                // Small optimization: don't have to load the class again since it was given to us in
-                // a register.
-                assert(rewrite_args->obj == obj_saved);
-            } else {
-                rewrite_args->obj = rewrite_args->rewriter->loadConst((intptr_t)base, Location::any());
-            }
-            val = base->getattr(attr, rewrite_args);
-            assert(rewrite_args->out_success);
-            if (val)
+        // Guard on the total shape of the class.
+        // from here we can assume that the index into the mro + the offset into the object remain the same.
+        Shape* total_shape = cls->computeTotalShape();
```

Comment:
this is probably low on the todo list, but maybe we should call `computeTotalShape` at some point during class initialization?  otherwise, it seems like every new class will fail the guards since the stored totalShape will be null.

also, seems like a decent time to remark that we should probably sort the attributes during typeNew -- they're just sorted by however std::unordered_map decides to iterate over them.